frontend/pipeline.yml

```
@@ -0,0 +1,300 @@
+        compare_to: '${RULES_CHANGES_COMPARE_TO_REF}'
+  needs: []
+  cache:
+    key:
+      files:
+        - package-lock.json
+    paths:
+      - frontend/.npm/
+  script:
+    - cd frontend
+    - npm ci --cache .npm --prefer-offline
+    - npm run test -- --ci --maxWorkers=2 --reporters=default --reporters=jest-junit
+  after_script:
+    - mkdir logs && cp .npm/_logs/* logs/
+  artifacts:
+    expire_in: 2w
+    paths:
+      - frontend/coverage
+      - frontend/logs
+    reports:
+      junit: junit.xml
+    when: always
+  tags:
+    - hetzner-amd-beefy
+
+test:frontend:docs-links:
+  stage: pre-test
+  image: ${CI_DEPENDENCY_PROXY_DIRECT_GROUP_IMAGE_PREFIX}/${NODE_IMAGE}
+  rules:
+    - if: $CI_COMMIT_REF_NAME !~ /^(staging|saas-[a-zA-Z0-9.]+)$/
+    - changes:
+        paths: ['frontend/**/*']
+        compare_to: '${RULES_CHANGES_COMPARE_TO_REF}'
+  needs: []
+  before_script:
+    - apk add --no-cache curl
+    - cd frontend
+  script:
+    - links=$(grep -r docs.mender.io src/ | grep -v snapshots | sed -e 's,\${docsVersion},'${DOCS_VERSION}'/,g' | sed -e 's,\${path},''/,g' | egrep -o 'https?://[^ `"]+' | sort | uniq)
+    - error=0
+    - |
+      for url in $links; do
+        if ! curl --output /dev/null --silent --head --fail $url; then
+          error=1
+          echo "[ko] $url"
+        else
+          echo "[ok] $url"
+        fi
+      done
+      if [ $error -gt 0 ]; then
+        exit 1
+      fi
+  tags:
+    - hetzner-amd-beefy
+
+test:frontend:docs-links:hosted:
+  extends: test:frontend:docs-links
+  rules:
+    - if: $CI_COMMIT_REF_NAME =~ /^(staging|saas-[a-zA-Z0-9.]+)$/
+  variables:
+    DOCS_VERSION: hosted
+
+build:frontend:docker:
+  extends: .template:build:docker
+  rules:
+    - changes:
+        paths: ['frontend/**/*']
+        compare_to: '${RULES_CHANGES_COMPARE_TO_REF}'
+      when: always
+  script:
+    - cd frontend
+    - docker context create ci;
+    - docker buildx create --name gui-builder --driver=docker-container ci
+    # needed for the license generation step
+    - docker build
+      --tag ${FRONTEND_REPOSITORY}:base
+      --target base
+      --builder=gui-builder
+      --output type=docker,dest=baseImage.tar
+      .
+    # build production target
+    - docker build
+      --tag ${CI_REGISTRY_IMAGE}:frontend-${CI_PIPELINE_ID}
+      --builder=gui-builder
+      --build-arg GIT_COMMIT_TAG="${CI_COMMIT_TAG}"
+      --platform ${DOCKER_PLATFORM:-linux/amd64}
+      --provenance false
+      --push
+      .
+  artifacts:
+    paths:
+      - baseImage.tar
+
+.template:test:frontend:acceptance:
+  stage: e2e-test
+  image: ${CI_DEPENDENCY_PROXY_DIRECT_GROUP_IMAGE_PREFIX}/tiangolo/docker-with-compose
+  rules:
+    - changes:
+        paths: ['frontend/**/*']
+        compare_to: '${RULES_CHANGES_COMPARE_TO_REF}'
+  needs:
+    - job: build:frontend:docker
+      artifacts: true
+  services:
+    - name: ${CI_DEPENDENCY_PROXY_DIRECT_GROUP_IMAGE_PREFIX}/docker:dind
```

Can we use a consistent version of docker?
```suggestion
    - name: ${CI_DEPENDENCY_PROXY_DIRECT_GROUP_IMAGE_PREFIX}/docker:${DOCKER_VERSION}-dind
```

---

considering that we have missed upgrading these and they won't be tracked through dependabot I would rather not.

---

That is exactly what I was intending to solve by introducing one version that gets reused instead of each job declaring their own image that is not looked after. Pulling the image through the cache may lead to stale versions of this tag and then you could have version skew between the Docker client and server.

---

Sure, OTOH we have had several pipelines that had it defined in the ci file and went stale, whereas others were diligently updated by hand and again others were left with the mutable tag and signaled problematic changes early... if you want to guarantee that we get this kept up to date, we can set it here as well - the state in the majority of the older repos says otherwise IMO.
As to the skewing: I don't think there will be version skew for any reasonable amount of time on this proxy and this image in particular.